### PR TITLE
[KYUUBI #2854] Add exception field in KyuubiSessionEvent

### LINKED ISSUE #2854: [Improvement] Add exception field in KyuubiSessionEvent
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

An exception may occur when opening session. I want to record it in KyuubiSessionEvent.

### How should we improve?

add exception field in KyuubiSessionEvent

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by wForget**
### _Why are the changes needed?_

close #2854 

### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala [hunk 1]
```diff
@@ -52,7 +52,8 @@ case class KyuubiSessionEvent(
     var engineId: String = "",
     var openedTime: Long = -1L,
     var endTime: Long = -1L,
-    var totalOperations: Int = 0) extends KyuubiEvent {
+    var totalOperations: Int = 0,
+    var exception: Option[Throwable] = None) extends KyuubiEvent {
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala:56]: The scaladoc is outdated

**Review by pan3793 — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala (hunk 1 above)
Comment (wForget) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala:57]: > scaladoc

thanks, updated

**Comment by wForget** [comment]
irrelevant failure
```
StatementSuite:
- test select *** FAILED ***
  java.sql.SQLException: org.apache.kyuubi.KyuubiSQLException: Error operating EXECUTE_STATEMENT: java.sql.SQLException: errCode = 2, detailMessage = Failed to find enough host with storage medium and tag(HDD/{"location" : "default"}) in all backends. need: 1
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.StatementImpl.executeInternal(StatementImpl.java:768)
	at com.mysql.cj.jdbc.StatementImpl.execute(StatementImpl.java:653)
	at org.apache.kyuubi.engine.jdbc.operation.ExecuteStatement.org$apache$kyuubi$engine$jdbc$operation$ExecuteStatement$$executeStatement(ExecuteStatement.scala:60)
	at org.apache.kyuubi.engine.jdbc.operation.ExecuteStatement$$anon$1.run(ExecuteStatement.scala:43)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
	at org.apache.kyuubi.KyuubiSQLException$.apply(KyuubiSQLException.scala:69)
	at org.apache.kyuubi.engine.jdbc.operation.JdbcOperation$$anonfun$onError$1.applyOrElse(JdbcOperation.scala:84)
	at org.apache.kyuubi.engine.jdbc.operation.JdbcOperation$$anonfun$onError$1.applyOrElse(JdbcOperation.scala:69)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at org.apache.kyuubi.engine.jdbc.operation.ExecuteStatement.org$apache$kyuubi$engine$jdbc$operation$ExecuteStatement$$executeStatement(ExecuteStatement.scala:88)
	at org.apache.kyuubi.engine.jdbc.operation.ExecuteStatement$$anon$1.run(ExecuteStatement.scala:43)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.sql.SQLException: errCode = 2, detailMessage = Failed to find enough host with storage medium and tag(HDD/{"location" : "default"}) in all backends. need: 1
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.StatementImpl.executeInternal(StatementImpl.java:768)
	at com.mysql.cj.jdbc.StatementImpl.execute(StatementImpl.java:653)
	at org.apache.kyuubi.engine.jdbc.operation.ExecuteStatement.org$apache$kyuubi$engine$jdbc$operation$ExecuteStatement$$executeStatement(ExecuteStatement.scala:60)
	... 6 more
  at org.apache.kyuubi.jdbc.hive.KyuubiStatement.waitForOperationToComplete(KyuubiStatement.java:354)
  at org.apache.kyuubi.jdbc.hive.KyuubiStatement.executeWithConfOverlay(KyuubiStatement.java:200)
  at org.apache.kyuubi.jdbc.hive.KyuubiStatement.execute(KyuubiStatement.java:194)
  at org.apache.kyuubi.engine.jdbc.doris.StatementSuite.$anonfun$new$2(StatementSuite.scala:29)
  at org.apache.kyuubi.engine.jdbc.doris.StatementSuite.$anonfun$new$2$adapted(StatementSuite.scala:24)
  at org.apache.kyuubi.operation.JDBCTestHelper.$anonfun$withMultipleConnectionJdbcStatement$3(JDBCTestHelper.scala:60)
  at org.apache.kyuubi.operation.JDBCTestHelper.$anonfun$withMultipleConnectionJdbcStatement$3$adapted(JDBCTestHelper.scala:60)
  at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
  at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
  ...
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSession.scala [hunk 2]
```diff
@@ -34,4 +34,14 @@ abstract class KyuubiSession(
...
+  private[kyuubi] def handleSessionException(f: => Unit): Unit = {
+    try {
+      f
+    } catch {
+      case t: Throwable =>
+        getSessionEvent.foreach(_.exception = Some(t))
+        throw t
+    }
+  }
+
```
Comment (cxzl25) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSession.scala:46]: nit: remove empty line L46

Comment (wForget) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSession.scala]: fixed, thanks

**Review by cxzl25 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merging to master

**Pull request closed without merging**